File: 2021/11/12/shift_grid.py
"""
https://leetcode.com/problems/shift-2d-grid/
Author: Andrew Jarombek
Date: 11/12/2021
"""

from typing import List


def shift_grid(grid: List[List[int]], k: int) -> List[List[int]]:
    row = len(grid)
    col = len(grid[0])
    moves = k % (row * col)

    result = [[0 for c in range(col)] for r in range(row)]

    for r in range(row):
        for c in range(col):
            location = (r * col) + c
            source_location = (location - moves) % (row * col)
            source_row = source_location // col
            source_col = source_location % col
            result[r][c] = grid[source_row][source_col]

    return result


if __name__ == '__main__':
    assert shift_grid([[1, 2, 3], [4, 5, 6], [7, 8, 9]], 1) == [[9, 1, 2], [3, 4, 5], [6, 7, 8]]

File: 2022/08/01/shiftGrid.js
/**
 * https://leetcode.com/problems/shift-2d-grid/
 * @author Andrew Jarombek
 * @since 8/1/2022
 */

const shiftGrid = function(grid, k) {
    const values = [];
    const result = [];

    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[i].length; j++) {
            values.push(grid[i][j]);
        }
    }

    for (let i = 0; i < grid.length; i++) {
        const row = [];

        for (let j = 0; j < grid[i].length; j++) {
            let location = (j + (i * grid[i].length)) - k;
            let index = ((location % values.length) + values.length) % values.length;
            row.push(values[index]);
        }

        result.push(row);
    }

    return result;
}

const result = shiftGrid([[1, 2, 3], [4, 5, 6], [7, 8, 9]], 1);
const expectedResult = [[9, 1, 2], [3, 4, 5], [6, 7, 8]]

console.assert(expectedResult.every(
    (row, i) => row.every(
        (value, j) => value ===result[i][j]
    )
));
